internal/testrunner/runners/pipeline/benchmark.go

```diff
@@ -0,0 +1,388 @@
+		}
+	}
+	return step, nil
+}
+
+type processorPerformance struct {
+	key   string
+	value int64
+}
+
+type aggregation struct {
+	result []processorPerformance
+	err    error
+}
+
+type (
+	keyFn     func(ingest.Pipeline, ingest.Processor) string
+	valueFn   func(record ingest.StatsRecord) int64
+	mapFn     func(processorPerformance) testrunner.BenchmarkValue
+	compareFn func(a, b processorPerformance) bool
+	filterFn  func(processorPerformance) bool
+)
+
+func (ir ingestResult) aggregate(key keyFn, value valueFn) (agg aggregation) {
+	pipelines := make(map[string]ingest.Pipeline, len(ir.pipelines))
+	for _, p := range ir.pipelines {
+		pipelines[p.Name] = p
+	}
+
+	for pipelineName, pipelineStats := range ir.stats {
+		pipeline, ok := pipelines[pipelineName]
+		if !ok {
+			return aggregation{err: fmt.Errorf("unexpected pipeline '%s'", pipelineName)}
+		}
+		processors, err := pipeline.Processors()
+		if err != nil {
+			return aggregation{err: err}
+		}
+		if nSrc, nStats := len(processors), len(pipelineStats.Processors); nSrc != nStats {
+			return aggregation{err: fmt.Errorf("pipeline '%s' processor count mismatch. source=%d stats=%d", pipelineName, nSrc, nStats)}
+		}
+		for procId, procStats := range pipelineStats.Processors {
+			agg.result = append(agg.result, processorPerformance{
+				key:   key(pipeline, processors[procId]),
+				value: value(procStats.Stats),
+			})
+		}
+	}
+	return agg
+}
+
+func (agg aggregation) sort(compare compareFn) aggregation {
+	if agg.err != nil {
+		return agg
+	}
+	sort.Slice(agg.result, func(i, j int) bool {
+		return compare(agg.result[i], agg.result[j])
+	})
+	return agg
+}
+
+func ascending(a, b processorPerformance) bool {
+	return a.value < b.value
+}
+
+func descending(a, b processorPerformance) bool {
+	return !ascending(a, b)
+}
+
+func (agg aggregation) top(n int) aggregation {
+	if n < len(agg.result) {
+		agg.result = agg.result[:n]
+	}
+	return agg
+}
+
+func (agg aggregation) filter(keep filterFn) aggregation {
+	if agg.err != nil {
+		return agg
+	}
+	o := 0
+	for _, entry := range agg.result {
+		if keep(entry) {
+			agg.result[o] = entry
+			o++
+		}
+	}
+	agg.result = agg.result[:o]
+	return agg
+}
+
+func (agg aggregation) collect(fn mapFn) ([]testrunner.BenchmarkValue, error) {
+	if agg.err != nil {
+		return nil, agg.err
+	}
+	r := make([]testrunner.BenchmarkValue, len(agg.result))
+	for idx := range r {
+		r[idx] = fn(agg.result[idx])
+	}
+	return r, nil
+}
+
+func runSingleBenchmark(options testrunner.TestOptions, docs []json.RawMessage) (ingestResult, error) {
+	if len(docs) == 0 {
+		return ingestResult{}, errors.New("no docs supplied for benchmark")
+	}
+	dataStreamPath, found, err := packages.FindDataStreamRootForPath(options.TestFolder.Path)
+	if err != nil {
+		return ingestResult{}, errors.Wrap(err, "locating data_stream root failed")
+	}
+	if !found {
+		return ingestResult{}, errors.New("data stream root not found")
+	}
+
+	testCase := testCase{
+		events: docs,
```

Comment:
This might be better as a feature request.

I'm wondering if we could introduce `batch_size` here, and divide the number of docs into those batches and send each batch with `simulatePipelineProcessing`.  That would get around the limitation we have right now wher the number of docs must fit in one request and would make the processing more like what the users sees in production.  Might even be possible to concurrently call `simulatePipelineProcessing` to again be more like what the users experiences in production.